### PR TITLE
Use latest peft/transformers/accelerate/bitsandbytes for 4-bit (qlora)

### PR DESCRIPTION
**Pull request by arnocandel**
https://arxiv.org/abs/2305.14314
https://huggingface.co/blog/4bit-transformers-bitsandbytes
- [x] update deps
- [x] enable 4-bit generation
- [x] enable 4-bit lora training

Addresses #136 more directly than #107, but only at runtime.

### CODE REVIEW

File: finetune.py [hunk 1]
```diff
@@ -65,7 +65,8 @@ def train(
         micro_batch_size: int = 4,
         gradient_checkpointing=False,  # unnecessary with gradient accumulation enabled
         fp16=True,
-        train_8bit=True,
+        train_8bit=False,
```
Comment (arnocandel) [inline finetune.py:68]: disable 8-bit training by default, since much slower than 16-bit.

File: generate.py [hunk 2]
```diff
@@ -270,10 +271,12 @@ def main(
                 base_model = 'h2oai/h2ogpt-oasst1-512-12b'
                 # don't set load_8bit if passed base_model, doesn't always work so can't just override
                 load_8bit = True
+                load_4bit = False  # FIXME - consider using 4-bit instead of 8-bit
```
Comment (arnocandel) [inline generate.py:274]: first need to do performance testing, for both speed & accuracy

**Comment by arnocandel** [comment]
Testing 65B on 2x A6000Ada:
`(env) arno@rippa:/nfs4/llm/h2ogpt(4bit)$ CUDA_VISIBLE_DEVICES=0,1 torchrun --nproc_per_node=2 finetune.py --base_model=decapoda-research/llama-65b-hf --train_4bit=True --micro_batch_size=1 --run_id=3 --data_path=h2oai/openassistant_oasst1_h2ogpt_graded`
works:
`  0%|▋                                                                                                                                                            | 1/237 [04:27<17:33:37, 267.87s/it]`
image

**Review by pseudotensor — APPROVED** [review]
We will do more full testing of eval on 4bit vs. 8bit vs. 16bit.  According to QLORA paper, as long as do LORA on all linear layers, then should be as good as 16bit.  I'm unclear on speed of new Tim stuff.
